### PR TITLE
[REEF-554] Update IMRU APIs to enable writing of REEF IMRU driver and client

### PR DESCRIPTION
**Pull request by dkm2110**
This addressed the issue by
  * modifying the APIs
  * modifying MapperCount example to be compatible with new API
  * modifying InProcessIMRU to be compatible with the current API

JIRA:
  [REEF-554](https://issues.apache.org/jira/browse/REEF-554)

### CODE REVIEW

**Comment by dkm2110** [comment]
@markusweimer , created pull request for APIs of IMRU

File: lang/cs/Org.Apache.REEF.IMRU.Tests/MapperCountTest.cs [hunk 1]
```diff
@@ -30,7 +30,7 @@ namespace Org.Apache.REEF.IMRU.Tests
     [TestClass]
     public class MapperCountTest
     {
-        private const int NumberOfMappers = 7;
+        private const int NumberOfMappers = 1;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU.Tests/MapperCountTest.cs]: Why this change? I deliberately choose a prime number here to unearth any potential assumptions in the code.

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs [hunk 2]
```diff
@@ -27,23 +27,139 @@ namespace Org.Apache.REEF.IMRU.API
...
+        /// <param name="updateFunctionConfiguration">Update function configuration</param>
+        /// <param name="mapOutputPipelineDataConverterConfiguration">Configuration of 
+        /// PipelineDataConverter for TMapOutput</param>
+        /// <param name="mapInputPipelineDataConverterConfiguration">Configuration of 
+        /// PipelineDataConverter for TMapInput</param>
+        /// <param name="partitionedDatasetConfiguration">Configuration of partitioned 
+        /// dataset</param>
+        /// <param name="numberOfMappers">Number of mappers</param>
+        /// <param name="jobName">Job name</param>
+        internal IMRUJobDefinition(
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: Why the split into many different configurations? Did the approach discussed via email (making the various Codecs named parameters) not work?

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs [hunk 3]
```diff
@@ -55,21 +56,77 @@ public IMRUJobDefinitionBuilder SetJobName(string name)
...
-            {
-                throw new NullReferenceException("Configuration can't be null.");
-            }
             if (null == _jobName)
             {
                 throw new NullReferenceException("JobName can't be null.");
             }
-            return new IMRUJobDefinition(_configuration, _jobName);
+
+            return new IMRUJobDefinition(
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: Please check for `null` on all required parameters and throw appropriate exceptions.

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs (hunk 2 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: So for groupcommunications to work, I have to give the union of all codecs
as the service configuration. There we need to make sure that there are no
duplications.

File: lang/cs/Org.Apache.REEF.IMRU/Org.Apache.REEF.IMRU.csproj [hunk 4]
```diff
@@ -91,6 +92,10 @@ under the License.
       <Project>{cdfb3464-4041-42b1-9271-83af24cd5008}</Project>
       <Name>Org.Apache.REEF.Wake</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Org.Apache.REEF.IO\Org.Apache.REEF.IO.csproj">
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/Org.Apache.REEF.IMRU.csproj]: use `SolutionDir` instead of `..`

File: lang/cs/Org.Apache.REEF.Network.Examples.Client/BroadcastAndReduceClient.cs [hunk 5]
```diff
@@ -43,7 +43,7 @@ public void RunBroadcastAndReduce(bool runOnYarn, int numTasks, int startingPort
             const string driverId = "BroadcastReduceDriver";
             const string groupName = "BroadcastReduceGroup";
             const string masterTaskId = "MasterTask";
-            const int fanOut = 2;
+            const int fanOut = 2;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network.Examples.Client/BroadcastAndReduceClient.cs]: this change is purely syntactic, please remove.

**Comment by markusweimer** [comment]
@dkm2110 I've done a pass.

**Comment by dkm2110** [comment]
Wow that was super quick :) .

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs (hunk 2 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: How would there be duplications, if the `Codec` implementations are bound to `NamedParameter`s?

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: Consider the case where both codecs of TMapInput and TMapOutput have some
base codec configuration which is same. For example, in lbfgs both
TMapinput and TMapOutput are going to have same vector codec configuration.
Isn't it going to create a conflict.

Dhruv

File: lang/cs/Org.Apache.REEF.Network.Examples.Client/BroadcastAndReduceClient.cs (hunk 5 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.Network.Examples.Client/BroadcastAndReduceClient.cs]: done

**Comment by dkm2110** [comment]
@markusweimer  I have taken another pass

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs (hunk 2 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: I don't think so. As long as we bind the same implementation to both named parameters, we should be good to go.

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: Markus,

Just to be sure if I have MapInput which is nothing but say consists of two
floats and hence in its MapInput codec config say config1  I would have
binded IStreamingCodec<float> to say FloatCodec as base codec and I have
MapOutput which is same as MapOutput and in its  codec config say config2 I
have again   binded IStreamingCodec<float> to FloatCodec, and then I do
Configurations.Merge(config1,config2), it will not throw me the error?

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: To be more specific, say the two named parameters correspond to binding of configuration of IStreaming<TMapInput> and IStreaming<TMapOutput>. but they will also need to append this with base configurations config1 and config2 below.

 var config1 = TangFactory.GetTang().NewConfigurationBuilder()
                .BindImplementation(GenericType<IStreamingCodec<float>>.Class, GenericType<FloatStreamingCodec>.Class)
                .Build();
            var config2 = TangFactory.GetTang().NewConfigurationBuilder()
                .BindImplementation(GenericType<IStreamingCodec<float>>.Class, GenericType<FloatStreamingCodec>.Class)
                .Build();

So when now we merge together there is going to be conflict.

Are you saying that we put the whole configuration of TMapInput codec including the base one as one named parameter and the one corresponding to TMapOutput as another one.

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: plus does it make a difference if user gives split job definition versus single one. Even in case of single one, he has to bind to different named parameters in any case.

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: Ah, I think I finally got your concern. The question is not whether we can bind the same codec implementation for `TMapInput` and `TMapOutput`. We can do that by using named parameters. Your concern is: What if this codec has additional named parameters, which have to be bound to different values? That indeed is impossible, in the absence of [REEF-31](https://issues.apache.org/jira/browse/REEF-31). 

Did I understand correctly?

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: Yes dot on :)

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: And i thought about it. it's going to happen frequently

**Comment by markusweimer** [comment]
Now that my confusion is resolved, I'm doing another pass.

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs [hunk 6]
```diff
@@ -27,23 +27,139 @@ namespace Org.Apache.REEF.IMRU.API
...
+        /// <summary>
+        /// Name of the job
+        /// </summary>
+        internal string JobName
         {
-            get { return _configuration; }
+            get { return _jobName; }
         }
 
-        internal string JobName
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: All the `Set...()` methods have been replaced with properties. What's the rationale for that? 

Background: The reason I used methods before was to allow for fluent chaining of calls, e.g. `.SetJobName("foo").setMapOutputCodecConfiguration(...).....Build()`.

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs [hunk 7]
```diff
@@ -55,21 +61,108 @@ public IMRUJobDefinitionBuilder SetJobName(string name)
...
-                throw new NullReferenceException("Configuration can't be null.");
+                Exceptions.CaughtAndThrow(new NullReferenceException("Job name cannot be null"),
+                                    Level.Error, Logger);
             }
-            if (null == _jobName)
+
+            if (null == MapFunctionConfiguration)
             {
-                throw new NullReferenceException("JobName can't be null.");
+                Exceptions.CaughtAndThrow(new NullReferenceException("Map function configuration cannot be null"),
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: All of these should just be `Exceptions.Throw()` right? We are not catching exceptions here.

File: lang/cs/Org.Apache.REEF.IMRU/Examples/MapperCount/MapperCount.cs [hunk 8]
```diff
@@ -39,29 +40,49 @@ private MapperCount(IIMRUClient<int, int, int> imruClient)
...
+            jobDefinitionBuilder.NumberOfMappers = numberofMappers;
+
+            var results = _imruClient.Submit(jobDefinitionBuilder.SetJobName("MapperCount").Build());
+
+            if (results != null)
+            {
+                return results.First();
+            }
+
+            return -1;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/Examples/MapperCount/MapperCount.cs]: You might want to add a log line here, as it essentially is a failed job.

File: lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs [hunk 9]
```diff
@@ -52,9 +54,32 @@ private InProcessIMRUClient([Parameter(typeof(NumberOfMappers))] int numberOfMap
...
+            var updateFunctionCodecConfig = jobDefinition.UpdateFunctionCodecsConfiguration;
+            var mapInputPipelineConfig = jobDefinition.MapInputPipelineDataConverterConfiguration;
+            var mapOutputPipelineConfig = jobDefinition.MapOutputPipelineDataConverterConfiguration;
+
+            if (typeof (TMapInput) == typeof (TMapOutput))
+            {
+                mapOutputPipelineConfig = emptyConfig;
+            }
+
+            var mergedConfig = Configurations.Merge(
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs]: Given the reasoning above, we should consider keeping them separate even in the local execution model. Maybe as a separate JIRA?

**Comment by markusweimer** [comment]
@dkm2110 Done with another pass.

**Comment by jwang98052** [comment]
I will look at it as well.

**Comment by markusweimer** [comment]
@jwang98052 Thanks! Please ping me when you need me again on this PR.

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs (hunk 6 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: I assume you are talking about IMRU JOB definition builder and not IMRU JObDeifniton since comment is under job definiton

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: IF yes, I will make the change

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs (hunk 7 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: will change

File: lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs (hunk 9 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs]: ok, I will do that as separate jira later.

File: lang/cs/Org.Apache.REEF.IMRU/Examples/MapperCount/MapperCount.cs (hunk 8 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/Examples/MapperCount/MapperCount.cs]: no thats not true...For REEF IMRU, results will be written directly on to hdfs or local files and will not be sent to the driver or client. Hence, I am assuming that for those cases it would be null.

**Comment by dkm2110** [comment]
@markusweimer I have incorporated your changes. @jwang98052 lemme know when you are done

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs [hunk 10]
```diff
@@ -28,19 +32,21 @@ namespace Org.Apache.REEF.IMRU.API
...
         /// </summary>
-        /// <param name="configuration">The Configuration used to instantiate the IMapFunction instance.</param>
-        /// <seealso cref="IMRUConfiguration{TMapInput,TMapOutput,TResult}" />
-        /// <returns>this</returns>
-        public IMRUJobDefinitionBuilder SetConfiguration(IConfiguration configuration)
+        public IMRUJobDefinitionBuilder()
         {
-            _configuration = configuration;
-            return this;
+            MapInputPipelineDataConverterConfiguration = EmptyConfiguration;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: Are you intent to share the same Configuration instance between those two?

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs [hunk 11]
```diff
@@ -55,21 +61,108 @@ public IMRUJobDefinitionBuilder SetJobName(string name)
...
+        /// </summary>
+        public IConfiguration ReduceFunctionConfiguration { get; set; }
+
+        /// <summary>
+        /// Configuration of update function
+        /// </summary>
+        public IConfiguration UpdateFunctionConfiguration { get; set; }
+
+        /// <summary>
+        /// Configuration of PipelineDataConverter for Map outout
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: typo

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs (hunk 7 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: That's right

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs [hunk 12]
```diff
@@ -55,21 +61,108 @@ public IMRUJobDefinitionBuilder SetJobName(string name)
...
+        /// TODO: automatically from IPartitionedDataset. However, right now 
+        /// TODO: GroupComm. instantiated in IMRUDriver needs this parameter 
+        /// TODO: in constructor. This will be removed once we remove it from GroupComm. 
+        public int NumberOfMappers { get; set; }
+
+        /// <summary>
         /// Instantiate the IMRUJobDefinition.
         /// </summary>
         /// <returns>The IMRUJobDefintion configured.</returns>
         /// <exception cref="NullReferenceException">If any of the required paremeters is not set.</exception>
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: typo

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUGenericTypePipelineDataConverterConfiguration.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+using Org.Apache.REEF.Tang.Util;
+
+namespace Org.Apache.REEF.IMRU.API
+{
+    /// <summary>
+    /// A configuration module for The PipelineDataConverter used for chunking 
+    /// and dechunking TMapInput and TMapOutput
+    /// </summary>
+    /// <typeparam name="T">Generic type</typeparam>
+    public sealed class IMRUGenericTypePipelineDataConverterConfiguration<T> : ConfigurationModuleBuilder
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUGenericTypePipelineDataConverterConfiguration.cs]: Why is called GenericType? The class name is too long.

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs [hunk 14]
```diff
@@ -55,21 +61,108 @@ public IMRUJobDefinitionBuilder SetJobName(string name)
...
+
+        /// <summary>
+        /// Configuration of update function
+        /// </summary>
+        public IConfiguration UpdateFunctionConfiguration { get; set; }
+
+        /// <summary>
+        /// Configuration of PipelineDataConverter for Map outout
+        /// </summary>
+        public IConfiguration MapOutputPipelineDataConverterConfiguration { get; set; }
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: Why not use build pattern instead of using properties?

File: lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs [hunk 15]
```diff
@@ -52,9 +54,32 @@ private InProcessIMRUClient([Parameter(typeof(NumberOfMappers))] int numberOfMap
...
+        /// <returns>The result of the job</returns>
         public IEnumerable<TResult> Submit(IMRUJobDefinition jobDefinition)
         {
-            var injector = TangFactory.GetTang().NewInjector(jobDefinition.Configuration);
+            var emptyConfig = TangFactory.GetTang().NewConfigurationBuilder().Build();
+            var updateFunctionCodecConfig = jobDefinition.UpdateFunctionCodecsConfiguration;
+            var mapInputPipelineConfig = jobDefinition.MapInputPipelineDataConverterConfiguration;
+            var mapOutputPipelineConfig = jobDefinition.MapOutputPipelineDataConverterConfiguration;
+
+            if (typeof (TMapInput) == typeof (TMapOutput))
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs]: remove space after typeof

File: lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs [hunk 16]
```diff
@@ -52,9 +54,32 @@ private InProcessIMRUClient([Parameter(typeof(NumberOfMappers))] int numberOfMap
...
+        /// <summary>
+        /// Submits the map job
+        /// </summary>
+        /// <param name="jobDefinition">Job definition given by the user</param>
+        /// <returns>The result of the job</returns>
         public IEnumerable<TResult> Submit(IMRUJobDefinition jobDefinition)
         {
-            var injector = TangFactory.GetTang().NewInjector(jobDefinition.Configuration);
+            var emptyConfig = TangFactory.GetTang().NewConfigurationBuilder().Build();
+            var updateFunctionCodecConfig = jobDefinition.UpdateFunctionCodecsConfiguration;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs]: Why not use jobDefinition.UpdateFunctionCodecsConfiguration directly in the merge code below instead of create a local variable? That would be consistent with other code.

File: lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs [hunk 17]
```diff
@@ -52,9 +54,32 @@ private InProcessIMRUClient([Parameter(typeof(NumberOfMappers))] int numberOfMap
...
+            var mapInputPipelineConfig = jobDefinition.MapInputPipelineDataConverterConfiguration;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs]: Same for those two.

File: lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs (hunk 9 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs]: Your mentioned the reason to have multiple configurations is because they cannot be merged. Do you have issue in the Merge here?

File: lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs [hunk 18]
```diff
@@ -52,9 +54,32 @@ private InProcessIMRUClient([Parameter(typeof(NumberOfMappers))] int numberOfMap
...
+            var mapOutputPipelineConfig = jobDefinition.MapOutputPipelineDataConverterConfiguration;
+
+            if (typeof (TMapInput) == typeof (TMapOutput))
+            {
+                mapOutputPipelineConfig = emptyConfig;
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs]: You can merge every one except mapOutputPipelineConfig. Then add logic 
  if (typeof(TMapInput) != typeof(TMapOutput)), then merge with mapOutputPipelineConfig.
Then you don't need to create an empty config to merge with nothing. Performance would be better.

**Comment by jwang98052** [comment]
I complete this pass.

File: lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs (hunk 9 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/InProcess/InProcessIMRUClient.cs]: Yes, there are chances of conflict here also. Let me try to modify.

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs (hunk 10 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: Only if its empty in which case it should not matter right?

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs (hunk 11 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: fixed

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs (hunk 12 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: fixed

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUGenericTypePipelineDataConverterConfiguration.cs (hunk 13 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUGenericTypePipelineDataConverterConfiguration.cs]: because it is used for both TMapInput and TMapOutput

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs (hunk 14 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: I changed it to that...I think thats what was suggested by markus

**Comment by dkm2110** [comment]
@markusweimer @jwang98052  I have pushed in changes suggested by Julia. I simplified InProcessIMRUCLient a lot since Codecs and Pipeline Converter configs are not needed at all. Now only Update function and reduce function config needs to be merged which should not have conflict.

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs (hunk 6 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs]: Yes, I meant in the builder. Sorry about that.

File: lang/cs/Org.Apache.REEF.IMRU/Examples/MapperCount/MapperCount.cs (hunk 8 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/Examples/MapperCount/MapperCount.cs]: OK. In a future version of this, we might want to return a pointer to the results, then.

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/Examples/MapperCount/MapperCount.cs]: yes

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUGenericTypePipelineDataConverterConfiguration.cs (hunk 13 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUGenericTypePipelineDataConverterConfiguration.cs]: Yes, but why such a long name? Maybe `IMRUPipelineDataConverterConfiguration`?

**Comment by markusweimer** [comment]
+1 on merging from me.

@jwang98052 ?

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUGenericTypePipelineDataConverterConfiguration.cs (hunk 13 above)
Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUGenericTypePipelineDataConverterConfiguration.cs]: Renames and pushed

File: lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs (hunk 10 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: If the purpose is to merge it with others, then sharing the same empty instance is OK. If you might add some other settings in the configuration respectively, then sharing may cause problem.

Comment (dkm2110) [inline lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinitionBuilder.cs]: That will not happen since if user wants to give something else he will
create new instance. Note that there is no get method. So it cannot be
appended.

Dhruv

**Comment by jwang98052** [comment]
Looks good to me.

**Comment by markusweimer** [comment]
@jwang98052 Thanks, I will merge.
